File: C++/115. Distinct Subsequences.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        int m=1e9+7;
        int n1 = s.size();
        int n2 = t.size();
        vector<vector<int>> vec(n2+1,vector<int>(n1+1));
        for(int j=0; j<=n1; j++) {
            vec[0][j] = 1;
        }
        for(int i=0; i<n2; i++) {
            for(int j=0; j<n1; j++) {
                if(t[i] == s[j]) {
                    vec[i+1][j+1] = (vec[i][j] + vec[i+1][j])%m;
                }
                else {
                    vec[i+1][j+1] = vec[i+1][j];
                }
            }
        } 
        return vec[n2][n1];
    }
};

File: JavaScript/115. Distinct Subsequences.js
/**
 * @param {string} s
 * @param {string} t
 * @return {number}
 */
var numDistinct = function(s, t) {
    var n1 = s.length;
    var n2 = t.length;
    var arr = new Array(n2+1).fill(0).map(() => new Array(n1+1).fill(0));
    for(var j=0; j<=n1; j++) {
        arr[0][j] = 1;
    }
    for(var i=0; i<n2; i++) {
        for(var j=0; j<n1; j++) {
            if(t[i] == s[j]) {
                arr[i+1][j+1] = arr[i][j] + arr[i+1][j];
            }
            else {
                arr[i+1][j+1] = arr[i+1][j];
            }
        }
    } 
    return arr[n2][n1];
};
